### PR TITLE
GPT-J rope optimization on xpu

### PR DESCRIPTION
**Pull request by cyita**
## Description

1. Use `torch.ops.torch_ipex.apply_rotary_embedding_two_qk`.
2. Move sin/cos computation -> GPTJModel forward to avoid generating in attn

TODO:

- [x] Test transformers==4.31, xpu
- [x] Test transformers==4.36, xpu
- [x] Test transformers==4.31, cpu
- [x] Test transformers==4.36, cpu

### 4. How to test?
- [ ] N/A
- [ ] Unit test
- [ ] Application test
- [ ] Document test
- [ ] ...

### 5. New dependencies

- [ ] New Python dependencies
       - Dependency1 
       - Dependency2
       - ...
- [ ] New Java/Scala dependencies and their license
       - Dependency1 and license1
       - Dependency2 and license2
       - ...

### CODE REVIEW

File: python/llm/src/bigdl/llm/transformers/models/utils.py [hunk 1]
```diff
@@ -153,11 +153,10 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids, model_family):
         k_embed = (k * cos) + (rotate_half(k) * sin)
         return q_embed, k_embed
     elif model_family == "gptj":
-        cos = torch.repeat_interleave(cos[:, :, None, :], 2, 3)
-        sin = torch.repeat_interleave(sin[:, :, None, :], 2, 3)
-        q_embed = (q * cos) + (rotate_every_two(q) * sin)
-        k_embed = (k * cos) + (rotate_every_two(k) * sin)
-        return q_embed, k_embed
+        torch.ops.torch_ipex.apply_rotary_embedding_two_qk(
```
Comment (qiuxin2012) [inline python/llm/src/bigdl/llm/transformers/models/utils.py:156]: Don't change this. This `apply_rotary_pos_emb` is used for both cpu and xpu.

File: python/llm/src/bigdl/llm/transformers/models/gptj.py [hunk 2]
```diff
@@ -100,28 +103,22 @@ def gptj_attention_forward(
...
+    sin, cos = rotary_emb
 
     if self.rotary_dim is not None:
         k_rot = key[:, :, :, : self.rotary_dim]
-        k_pass = key[:, :, :, self.rotary_dim:]
-
         q_rot = query[:, :, :, : self.rotary_dim]
-        q_pass = query[:, :, :, self.rotary_dim:]
-
         q_rot, k_rot = apply_rotary_pos_emb(q_rot, k_rot, cos, sin, position_ids, "gptj")
```
Comment (qiuxin2012) [inline python/llm/src/bigdl/llm/transformers/models/gptj.py:121]: Add a comment for this? `torch.ops.torch_ipex.apply_rotary_embedding_two_qk` is inplacing.

File: python/llm/src/bigdl/llm/transformers/models/gptj.py [hunk 3]
```diff
@@ -343,6 +343,10 @@ def gptj_model_forward(
     all_self_attentions = () if output_attentions else None
     all_hidden_states = () if output_hidden_states else None
 
+    use_fuse_rope = input_ids.device.type == "xpu" and not self.training
```
Comment (qiuxin2012) [inline python/llm/src/bigdl/llm/transformers/models/gptj.py:346]: Move this to attention forward

**Comment by cyita** [comment]
ipex 2.0 : 17.7ms/token

File: python/llm/src/bigdl/llm/transformers/models/gptj.py (hunk 2 above)
Comment (cyita) [reply python/llm/src/bigdl/llm/transformers/models/gptj.py]: updated.

File: python/llm/src/bigdl/llm/transformers/models/utils.py (hunk 1 above)
Comment (cyita) [reply python/llm/src/bigdl/llm/transformers/models/utils.py]: Divided the cpu logic.

File: python/llm/src/bigdl/llm/transformers/models/gptj.py [hunk 4]
```diff
@@ -100,30 +104,44 @@ def gptj_attention_forward(
...
+            q_pass = query[:, :, :, self.rotary_dim:]
+            q_rot, k_rot = apply_rotary_pos_emb(q_rot, k_rot, cos, sin, position_ids, "gptj")
+            key = torch.cat([k_rot, k_pass], dim=-1)
+            query = torch.cat([q_rot, q_pass], dim=-1)
     else:
-        query, key = apply_rotary_pos_emb(query, key, cos, sin, position_ids, "gptj")
+        if use_fuse_rope:
+            # ipex's apply_rotary_embedding_two_qk can change the origin storage,
+            # so q/k will get the result directly.
+            from bigdl.llm.transformers.utils import get_ipex_version
```
Comment (qiuxin2012) [inline python/llm/src/bigdl/llm/transformers/models/gptj.py:135]: put this into a function

**Pull request merged**